src/connection_pool.rs

```diff
@@ -14,69 +14,90 @@ use tokio::sync::RwLock;
 // Pool for keeping open connections. Pooled connections are associated with a `ConnectionRemover`
 // which can be used to remove them from the pool.
 #[derive(Clone)]
-pub(crate) struct ConnectionPool {
-    store: Arc<RwLock<Store>>,
+pub(crate) struct ConnectionPool<T: Id> {
+    store: Arc<RwLock<Store<T>>>,
 }
 
-impl ConnectionPool {
+impl<T: Id> ConnectionPool<T> {
     pub fn new() -> Self {
         Self {
             store: Arc::new(RwLock::new(Store::default())),
         }
     }
 
-    pub async fn insert(&self, addr: SocketAddr, conn: quinn::Connection) -> ConnectionRemover {
+    pub async fn insert(
+        &self,
+        id: T,
+        addr: SocketAddr,
+        conn: quinn::Connection,
+    ) -> ConnectionRemover<T> {
         let mut store = self.store.write().await;
 
         let key = Key {
             addr,
             id: store.id_gen.next(),
         };
-        let _ = store.map.insert(key, conn);
+        let _ = store.id_map.insert(id, (conn.clone(), key));
+        let _ = store.key_map.insert(key, (conn, id));
 
         ConnectionRemover {
             store: self.store.clone(),
+            id,
             key,
         }
     }
 
-    pub async fn has(&self, addr: &SocketAddr) -> bool {
+    pub async fn has_addr(&self, addr: &SocketAddr) -> bool {
         let store = self.store.read().await;
 
         // Efficiently fetch the first entry whose key is equal to `key` and check if it exists
         store
-            .map
+            .key_map
             .range(Key::min(*addr)..=Key::max(*addr))
             .next()
             .is_some()
     }
 
+    #[allow(unused)]
+    pub async fn has_id(&self, id: &T) -> bool {
```

Comment:
perhaps we should rename it to `has_conn_id` ? and also perhaps have the methods like this one to expect an arg of type `&dyn ConnId` rather than having a generic on the whole structs?